### PR TITLE
Fix bitmap scan cost function c3_dBitmapPageCost

### PR DESCRIPTION
**Pull request by dgkimura**
Commit https://github.com/greenplum-db/gpdb/commit/fe7f1cb718e9958ea384e74c9db838a87c670a55 allowed ORCA to support bitmap index scans on mixed
heap/AO partitioned tables by circumstantially treating a btree index as
a bitmap index. However, ORCA requires the actual index type in order to
properly cost bitmap scan (e.g. bitmap index has additional bitmap page
cost as opposed to btree which does not).

Issue was identified through following setup:

```sql
CREATE TABLE test_partitioned (a integer not null, b text, c text)
WITH (appendonly=true) DISTRIBUTED BY (b)
PARTITION BY RANGE (a) (
    PARTITION p_pos START(0) END (1000000001),
    PARTITION p_neg START(-1000000000) END (0));
CREATE INDEX idx_test_partitioned ON test_partitioned(a);

CREATE TABLE batch (LIKE test_partitioned) DISTRIBUTED BY (b);

INSERT INTO test_partitioned SELECT a, a::text, a::text FROM generate_series(1, 10000000) a;
INSERT INTO test_partitioned SELECT -1 * a, a::text, a::text FROM generate_series(1, 10000000) a;
INSERT INTO batch SELECT * FROM test_partitioned WHERE random() < .00001;

ANALYZE batch; ANALYZE test_partitioned;

EXPLAIN SELECT count() FROM test_partitioned a WHERE EXISTS (SELECT 1 FROM batch b WHERE a.a = b.a);
```

Prior to this commmit ORCA incorrectly added bitmap page cost for a
btree index, thus over costing bitmap scan and picking seq scan plan
instead.

### CODE REVIEW

**Review by chrishajas — APPROVED** [review]
Looking at fe7f1cb again, it doesn't look like that impacts this PR. Rather, this piece of code that only marks the index as btree if it is a heap table (and bitmap otherwise) should be causing it. Your fix looks good and is addressing the root of the problem.

```
		else if (RELSTORAGE_HEAP == relstorage && BTREE_AM_OID == indRel->rd_rel->relam)
		{
			/*
			 * we only send btree indexes as type btree if it is a normal heap table,
			 * AO and AOCO tables with btree indexes are sent as type bitmap
			 */
			indType = INDTYPE_BTREE;
		}
		
		```

Also, we don't need this in master as we look specifically at the partitioned index (the underlying index) instead of building a logical index info structure that returned the bitmap index for AO tables previously. Does that sound right?

**Comment by dgkimura** [comment]
On Monday, June 13, 2022 2:12 PM Chris Hajas wrote:
> Looking at https://github.com/greenplum-db/gpdb/commit/fe7f1cb718e9958ea384e74c9db838a87c670a55 again, it doesn't look like that impacts this PR. Rather, this piece of code that only marks the index as btree if it is a heap table (and bitmap otherwise) should be causing it.

Ah, yeah.  Thanks for that correction!  I'll fix up the commit message accordingly

> Also, we don't need this in master as we look specifically at the partitioned index (the underlying index) instead of building a logical index info structure that returned the bitmap index for AO tables previously. Does that sound right?

Correct, it's not needed in master.  And yes, your statement sounds accurate to me.  In master we obtain indexes directly from the catalog (using code path as for non-partitioned tables).

File: src/backend/gporca/libnaucrates/include/naucrates/md/CMDIndexGPDB.h [hunk 1]
```diff
@@ -52,6 +52,9 @@ class CMDIndexGPDB : public IMDIndex
 	// index type
 	EmdindexType m_index_type;
 
+	// index physical type
+	EmdindexType m_index_physical_type;
+
```
Comment (gpopt) [inline src/backend/gporca/libnaucrates/include/naucrates/md/CMDIndexGPDB.h:57]: Can you add a comment to explain the difference between `m_index_type` and `m_index_physical_type`?

File: src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp [hunk 2]
```diff
@@ -75,6 +75,25 @@ static const ULONG cmp_type_mappings[][2] = {
 	{IMDType::EcmptL, CmptLT},	  {IMDType::EcmptG, CmptGT},
 	{IMDType::EcmptGEq, CmptGEq}, {IMDType::EcmptLEq, CmptLEq}};
 
+static IMDIndex::EmdindexType
+GetIndexTypeFromOid(OID index_oid)
```
Comment (gpopt) [inline src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp:79]: Does it return physical index type?
Can we add a comment?

**Review by gpopt — APPROVED** [review]
LGTM

File: src/backend/gporca/libnaucrates/include/naucrates/md/CMDIndexGPDB.h (hunk 1 above)
Comment (dgkimura) [reply src/backend/gporca/libnaucrates/include/naucrates/md/CMDIndexGPDB.h:59]: Added comment in 6416a9d1f785ac9e835ee8e511f3ae313d874ac9.

File: src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp (hunk 2 above)
Comment (dgkimura) [reply src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp:87]: Yeah, it should return physical index type. Added comment in 6416a9d1f785ac9e835ee8e511f3ae313d874ac9.

**Comment by dgkimura** [comment]
Merged. Thanks for reviews!

**Pull request merged**